### PR TITLE
Limit max export size

### PR DESCRIPTION
**Pull request by dominikl**
Corresponding Insight PR to  #4262
Limits the export as jpg/png/tif function to a certain maximum image size.

Test: See instruction on above mentioned PR

### CODE REVIEW

**Comment by pwalczysko** [comment]
According to the https://github.com/openmicroscopy/openmicroscopy/pull/4262 the limit is at 12k x 12k. But the export as jpeg of 8k x 8k.jpeg just failed for me, as the output says "Cannot export "Big" image as jpeg." So there is another limit in Insight ?

**Comment by pwalczysko** [comment]
A mulitselection of images followed by Export as jpeg ends up with

```
WARNING:omero.gateway:ApiUsageException on <class 'omero.gateway.OmeroGatewaySafeCallWrapper'> to <8620585e-f0b7-47b0-9eb8-04abcb3bf22eomero.api.RenderingEngine> getPixels((<ServiceOptsDict: {'omero.session.uuid': '7c0c4d04-b68c-42f2-9bfb-5048f8977988', 'omero.client.uuid': '865b7db5-b251-42d3-a246-7f35a1810aa2'}>,), {})
Traceback (most recent call last):
  File "/opt/hudson/workspace/OMERO-DEV-merge-deploy/src/dist/lib/python/omero/gateway/__init__.py", line 4170, in __call__
    return self.f(*args, **kwargs)
  File "/opt/hudson/workspace/OMERO-DEV-merge-deploy/src/dist/lib/python/omero_api_RenderingEngine_ice.py", line 475, in getPixels
    return _M_omero.api.RenderingEngine._op_getPixels.invoke(self, ((), _ctx))
ApiUsageException: exception ::omero::ApiUsageException
{
    serverStackTrace = ome.conditions.ApiUsageException: RenderingEngine not ready: Pixels object not set.
	at ome.services.RenderingBean.errorIfNullPixels(RenderingBean.java:1608)
	at ome.services.RenderingBean.getPixels(RenderingBean.java:1129)
	at sun.reflect.GeneratedMethodAccessor1922.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
	at ome.services.util.ServiceHandler.invoke(ServiceHandler.java:121)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
	at com.sun.proxy.$Proxy107.getPixels(Unknown Source)
	at sun.reflect.GeneratedMethodAccessor1922.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
	at ome.security.basic.BasicSecurityWiring.invoke(BasicSecurityWiring.java:93)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at ome.services.blitz.fire.AopContextInitializer.invoke(AopContextInitializer.java:43)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
	at com.sun.proxy.$Proxy107.getPixels(Unknown Source)
	at sun.reflect.GeneratedMethodAccessor1967.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at ome.services.blitz.util.IceMethodInvoker.invoke(IceMethodInvoker.java:172)
	at ome.services.throttling.Callback.run(Callback.java:56)
	at ome.services.throttling.InThreadThrottlingStrategy.callInvokerOnRawArgs(InThreadThrottlingStrategy.java:56)
	at ome.services.blitz.impl.AbstractAmdServant.callInvokerOnRawArgs(AbstractAmdServant.java:140)
	at ome.services.blitz.impl.RenderingEngineI.getPixels_async(RenderingEngineI.java:201)
	at sun.reflect.GeneratedMethodAccessor1966.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
	at omero.cmd.CallContext.invoke(CallContext.java:78)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
	at com.sun.proxy.$Proxy108.getPixels_async(Unknown Source)
	at omero.api._RenderingEngineTie.getPixels_async(_RenderingEngineTie.java:186)
	at omero.api._RenderingEngineDisp.___getPixels(_RenderingEngineDisp.java:838)
	at omero.api._RenderingEngineDisp.__dispatch(_RenderingEngineDisp.java:1571)
	at IceInternal.Incoming.invoke(Incoming.java:222)
	at Ice.ConnectionI.invokeAll(ConnectionI.java:2482)
	at Ice.ConnectionI.dispatch(ConnectionI.java:1258)
	at Ice.ConnectionI.message(ConnectionI.java:1213)
	at IceInternal.ThreadPool.run(ThreadPool.java:321)
	at IceInternal.ThreadPool.access$300(ThreadPool.java:12)
	at IceInternal.ThreadPool$EventHandlerThread.run(ThreadPool.java:693)
	at java.lang.Thread.run(Thread.java:745)

    serverExceptionClass = ome.conditions.ApiUsageException
    message = RenderingEngine not ready: Pixels object not set.
}
Traceback (most recent call last):
  File "./script", line 619, in <module>
    runScript()
  File "./script", line 604, in runScript
    fileAnnotation, message = batchImageExport(conn, scriptParams)
  File "./script", line 436, in batchImageExport
    for ch in img.getChannels():
  File "/opt/hudson/workspace/OMERO-DEV-merge-deploy/src/dist/lib/python/omero/gateway/__init__.py", line 7530, in getChannels
    self._conn.SERVICE_OPTS).iterateChannels())]
  File "/opt/hudson/workspace/OMERO-DEV-merge-deploy/src/dist/lib/python/omero/gateway/__init__.py", line 4173, in __call__
    return self.handle_exception(e, *args, **kwargs)
  File "/opt/hudson/workspace/OMERO-DEV-merge-deploy/src/dist/lib/python/omero/gateway/__init__.py", line 4170, in __call__
    return self.f(*args, **kwargs)
  File "/opt/hudson/workspace/OMERO-DEV-merge-deploy/src/dist/lib/python/omero_api_RenderingEngine_ice.py", line 475, in getPixels
    return _M_omero.api.RenderingEngine._op_getPixels.invoke(self, ((), _ctx))
omero.ApiUsageException: exception ::omero::ApiUsageException
{
    serverStackTrace = ome.conditions.ApiUsageException: RenderingEngine not ready: Pixels object not set.
	at ome.services.RenderingBean.errorIfNullPixels(RenderingBean.java:1608)
	at ome.services.RenderingBean.getPixels(RenderingBean.java:1129)
	at sun.reflect.GeneratedMethodAccessor1922.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
	at ome.services.util.ServiceHandler.invoke(ServiceHandler.java:121)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
	at com.sun.proxy.$Proxy107.getPixels(Unknown Source)
	at sun.reflect.GeneratedMethodAccessor1922.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
	at ome.security.basic.BasicSecurityWiring.invoke(BasicSecurityWiring.java:93)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at ome.services.blitz.fire.AopContextInitializer.invoke(AopContextInitializer.java:43)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
	at com.sun.proxy.$Proxy107.getPixels(Unknown Source)
	at sun.reflect.GeneratedMethodAccessor1967.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at ome.services.blitz.util.IceMethodInvoker.invoke(IceMethodInvoker.java:172)
	at ome.services.throttling.Callback.run(Callback.java:56)
	at ome.services.throttling.InThreadThrottlingStrategy.callInvokerOnRawArgs(InThreadThrottlingStrategy.java:56)
	at ome.services.blitz.impl.AbstractAmdServant.callInvokerOnRawArgs(AbstractAmdServant.java:140)
	at ome.services.blitz.impl.RenderingEngineI.getPixels_async(RenderingEngineI.java:201)
	at sun.reflect.GeneratedMethodAccessor1966.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
	at omero.cmd.CallContext.invoke(CallContext.java:78)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
	at com.sun.proxy.$Proxy108.getPixels_async(Unknown Source)
	at omero.api._RenderingEngineTie.getPixels_async(_RenderingEngineTie.java:186)
	at omero.api._RenderingEngineDisp.___getPixels(_RenderingEngineDisp.java:838)
	at omero.api._RenderingEngineDisp.__dispatch(_RenderingEngineDisp.java:1571)
	at IceInternal.Incoming.invoke(Incoming.java:222)
	at Ice.ConnectionI.invokeAll(ConnectionI.java:2482)
	at Ice.ConnectionI.dispatch(ConnectionI.java:1258)
	at Ice.ConnectionI.message(ConnectionI.java:1213)
	at IceInternal.ThreadPool.run(ThreadPool.java:321)
	at IceInternal.ThreadPool.access$300(ThreadPool.java:12)
	at IceInternal.ThreadPool$EventHandlerThread.run(ThreadPool.java:693)
	at java.lang.Thread.run(Thread.java:745)

    serverExceptionClass = ome.conditions.ApiUsageException
    message = RenderingEngine not ready: Pixels object not set.
}
```

**Comment by dominikl** [comment]
The error with the multiselection seems to be caused by the export script not by Insight. The export limitation to non-big images also seems to be server side. I'm giving up on this.

**Comment by jburel** [comment]
@pwalczysko @dominikl 
insight used the script to export as jpeg/png.
The script will need to be reviewed in order to support the changes.

**Comment by dominikl** [comment]
That means if @pwalczysko has checked the functionality with respect to the menus (if they are enabled/disabled when they are supposed to be), this PR could be merged, right? As it only deals with  handling the `omero.client.download_as.max_size` property in Insight. The export itself (and it's failing) is a matter of the export script.

**Comment by jburel** [comment]
@dominikl: for the export part (right-hand pane) yes.
But you also need to remove the check in the viewer i.e. Save As. (SaveAction)

**Comment by dominikl** [comment]
Ah thanks, another menu item I've not noticed before. Should be fixed with the last commit.

**Comment by jburel** [comment]
I tried to Save as jpeg a big image from main viewer. (user-1 gatan file)
```
java.lang.Exception: java.lang.NullPointerException
	at org.openmicroscopy.shoola.agents.imviewer.util.saver.ImgSaver.writeSingleImage(ImgSaver.java:222)
	at org.openmicroscopy.shoola.agents.imviewer.util.saver.ImgSaver.saveImage(ImgSaver.java:488)
	at
````

**Comment by jburel** [comment]
To test the last set of commits:
 * Open a big image in the full viewer e.g. gatan file
 * Go to Controls>Save As
 * Save the image as png for example (extension is not key!)

File: components/insight/SRC/org/openmicroscopy/shoola/agents/imviewer/browser/BrowserModel.java [hunk 1]
```diff
@@ -44,16 +45,22 @@
 import org.openmicroscopy.shoola.agents.imviewer.view.ImViewer;
 import org.openmicroscopy.shoola.agents.imviewer.view.ImViewerFactory;
 import org.openmicroscopy.shoola.agents.imviewer.view.ViewerPreferences;
+
 import ome.model.units.BigResult;
 import omero.log.LogMessage;
+
 import org.openmicroscopy.shoola.env.LookupNames;
+
 import omero.model.Length;
```
Comment (mtbc) [inline components/insight/SRC/org/openmicroscopy/shoola/agents/imviewer/browser/BrowserModel.java]: while fiddling with these, may want to keep the org ones together and the omero ones together?

**Comment by pwalczysko** [comment]
All good here. the test https://github.com/openmicroscopy/openmicroscopy/pull/4270#issuecomment-149845486 passes fully for big (even very big, 80 000 x 100 000) images.  The only problem in the ``Save as `` workflow seems to be the multiselection reported in https://github.com/ome/scripts/pull/105#issuecomment-150168171

**Comment by jburel** [comment]
I have modified insight and the script. so both should now work

**Comment by jburel** [comment]
@pwalczysko: note that the issue fixed by the last set of commits is not related to big images but to multi-selection.

**Comment by pwalczysko** [comment]
Note: the script itself has a problem - need to wait for the fix in https://github.com/ome/scripts/pull/105#issuecomment-150533523

**Comment by jburel** [comment]
@pwalczysko: script should be fixed now.

**Comment by pwalczysko** [comment]
Tested:
  - gatan from full viewer (export as all formats available)
  - gatan in a group of 4 images from the ``Save as jpeg``, ``Save as png``, ``Save as tiff``
  - leica-scn image (too big, over the 12x12 limit) in a multiselection with other images on ``Save as jpeg`` -> the too big image was left out, the other ones were exported as expected
  - single selections and ``Save as jpeg``

All passed. Ready to merge.

**Comment by jburel** [comment]
Thanks all

**Pull request merged**